base/meta.jl

```diff
@@ -73,10 +73,14 @@ julia> Meta.isexpr(ex, :call, 2)
 true
 ```
 """
-isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && ex.head === head
+function isexpr end
+
+# seemingly useless `head′` and `n′` are here just to help inference back-propagate
+# `isa(ex, Expr)` constraints to the caller inter-procedurally
+isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && (head′ = head; ex.head === head′)
 isexpr(@nospecialize(ex), heads) = isa(ex, Expr) && in(ex.head, heads)
-isexpr(@nospecialize(ex), head::Symbol, n::Int) = isa(ex, Expr) && ex.head === head && length(ex.args) == n
-isexpr(@nospecialize(ex), heads, n::Int) = isa(ex, Expr) && in(ex.head, heads) && length(ex.args) == n
+isexpr(@nospecialize(ex), head::Symbol, n::Int) = isa(ex, Expr) && (head′ = head; ex.head === head′) && (n′ = n; length(ex.args) == n′)
+isexpr(@nospecialize(ex), heads, n::Int) = isa(ex, Expr) && in(ex.head, heads) && (n′ = n; length(ex.args) == n′)
```

Comment:
IIUC, I think this problem should be solved now?

Reply:
Ah, I see, it still prioritizes the test against `head`, though it yields no beneficial information

Reply:
yeah, maybe this comment will explain to those who're interested in the future:
https://github.com/JuliaLang/julia/blob/b08f86f335a5a878e50a22714ae9233c5877f096/test/compiler/inference.jl#L1789-L1797

Reply:
I think fixing that is the last piece to solve for this PR (and would also fix the `test_const_return((x)->isdefined(x, :re), Tuple{ComplexF64}, true)` test). Namely, that Conditional results which don't constrain the input argument should be left as Const instead. Then, logic similar to what you wrote to handle ReturnNode, where you pick up the first "interesting" slot, would be needed in `abstract_call_gf_by_type` (when `napplicable > 1`) to form it just in time when it is needed.

ref https://github.com/JuliaLang/julia/pull/38905/files#r552171097

Reply:
Hmmm, do you mean something like [this](https://github.com/aviatesk/julia/compare/backprop2...aviatesk:backprop4) ?
I'm not sure I'm following your idea, but to me the diff looks a bit complicated and I feel it might be better to keep doing the "pick up interesting slot, covert boolean into conditional" work in just one place, for now.
Could you tell me why we should return `Const(::Bool)` instead of `InterConditional` which contains essential information about the constantness of its boolean value ?

I also wonder that it won't solve the original `isexpr` case even if we stop converting `Const(::Bool)` into `Conditional` in a callee frame.
As far as I understand, the current limitation is that we can't back-propagate multiple conditionals; as shown below,
with the lowering code obtained from the original `isexpr` definition, we will first see `Conditional(SlotNumber(3), Symbol, Symbol)` return type and so we will [try to convert the succeeding boolean return types into conditional on the `SlotNumber(3)`](https://github.com/JuliaLang/julia/blob/4fa55985685d45d8df77286c8c1bb927ca80a968/base/compiler/abstractinterpretation.jl#L1381-L1385), which is less interesting than that on `SlotNumber(2)`. In other word, since `SlotNumber(3)` is a valid argument and we first see it as return type, the current logic ends up missing the constraint of `SlotNumber(2)`.
This problem won't change whatever we try to convert `Const(false)` return type at `%7` into `Conditional`, no ?
```julia
julia> isexpr(@nospecialize(ex), head::Symbol) = isa(ex, Expr) && ex.head === head
isexpr (generic function with 1 method)

julia> code_lowered(isexpr, (Any,Symbol)) |> first
CodeInfo(
1 ─      nothing
│   %2 = ex isa Main.Expr
└──      goto #3 if not %2
2 ─ %4 = Base.getproperty(ex, :head)
│   %5 = %4 === head  # Conditional(SlotNumber(3), Symbol, Symbol)
└──      return %5    # Conditional(SlotNumber(3), Symbol, Symbol) -> Conditional(SlotNumber(3), Symbol, Symbol)
3 ─      return false # Const(false) -> Conditional(SlotNumber(3), Union{}, Symbol)
)
```
The new definition of `isexpr` setups a dummy local variable for `SlotNumber(3)` and prevents the logic to propagate its constraint (because the dummy is not a call argument, the logic will propagate the constraint of `SlotNumber(2)` instead).

I guess the best way to address this would be to setup new lattice element that contains multiple conditional constraints, back-propagate it and update states based with it. Since it will end up refactoring huge part of our lattice implementation, I thought it's better to be done in another PR.

Reply:
@vtjnash could I get a quick comment on this ? I'd like to know whether I'm understanding your idea or missing something.

Reply:
Yes, that commit looks about right. I ran into an unrelated bug trying to use it, so let me fix that first (https://github.com/JuliaLang/julia/issues/39218) and I'll push some updates

Reply:
okay, I'm being pretty busy for day work until this weekend, but will try to help you fix the bug once I find a time.

Reply:
Finished now in cb07897d87dd2b39019336e93c1d5586465277ef. WDYT?

Reply:
Thanks, #39221 looked good to me.

I just rebased this PR against the current `master`; I found some [tests with logging macro](https://github.com/aviatesk/julia/blob/18f5f7cef783acd9e00d7174023b1102afeea226/test/corelogging.jl#L97-L112) are failing on this PR somehow, and so will investigate into it, but I guess you may have better insight on this because the blame says you added those tests ?

I also updated [`backprop4`](https://github.com/aviatesk/julia/compare/backprop2...aviatesk:backprop4) branch, which tries to implement [your idea](https://github.com/JuliaLang/julia/pull/38905#discussion_r552353609). How you do want to review it ? Shall I push the commit into this PR, or craft a separate PR ?

Reply:
Okay I've tracked down the logging macro failures are because of a bug in this PR:
```julia
julia> code_typed((Any,); optimize = false) do x
           Base.CoreLogging.issimple(x) && return x
           nothing
       end |> first
CodeInfo(
1 ─ %1 = Base.CoreLogging::Core.Const(Base.CoreLogging)
│   %2 = Base.getproperty(%1, :issimple)::Core.Const(Base.CoreLogging.issimple)
│   %3 = (%2)(x)::Bool
└──      goto #3 if not %3
2 ─      return x::String
3 ─      return Main.nothing
) => Union{Nothing, String}
```
, and this leads to the wrong code generation of `logmsg`.

Reply:
So this PR is failing because e1859a1c25a returns `true` for different constant `Conditional`.
```julia
t1, t2 = InterConditional(2, Int, Union{}), InterConditional(2, Symbol, Union{}) # different conditionals
t1 ⊑ t2          # => true
tchanged(t1, t2) # => false
```
, with which we won't update `InterConditional` return type when there're multiple constant `Conditional` returns, e.g.
```julia
f(x) = begin
    isa(x, Symbol) && return true
    isa(x, Int) && return true
    false
end

code_typed((Any,); optimize = false) do x
    f(x) && return x
    nothing
end |> first |> last # => Union{Nothing, Symbol}
```

---

I think we can just fix this by not converting `Const(::Bool)` return type to `InterConditional`. Let me merge `backprop4` branch into this PR, and see how CI goes.

Reply:
I also added more fix for e1859a1 in d30287f; we need to unwrap same constant conditionals on different slots earlier, otherwise 
```julia
# before
Core.Compiler.tmerge(Core.Compiler.InterConditional(1, Int, Union{}), Core.Compiler.InterConditional(2, String, Union{})) # => Bool

# after
Core.Compiler.tmerge(Core.Compiler.InterConditional(1, Int, Union{}), Core.Compiler.InterConditional(2, String, Union{})) # => Const(true)
```

Reply:
Ah, sorry, I've reverted that bad commit.

Did you see my link https://github.com/JuliaLang/julia/commit/cb07897d87dd2b39019336e93c1d5586465277ef? That was based on your backprop4 branch, but should finish the implementation of it. You're welcome to take that and squash it into your commits. It should always handle the case you are concerned about there (though I don't currently have the branch checked out to verify for certain).

Reply:
Ah, sorry I've totally missed cb07897 ! 

The code looks good (rather, amazing) to me; I hand-copied your commit with some minor refactoring.
I will leave inline comments on where I modified a bit, but it should almost follow your implementation.

Thanks a lot for your help !